pkg/reconciler/service/controller.go

```diff
@@ -0,0 +1,143 @@
+	serviceinformer "knative.dev/eventing-istio/pkg/client/injection/kube/informers/core/v1/service"
+	"knative.dev/eventing-istio/pkg/client/injection/kube/reconciler/core/v1/service"
+	istioclientset "knative.dev/eventing-istio/pkg/client/istio/injection/client"
+	istionetworkinginformer "knative.dev/eventing-istio/pkg/client/istio/injection/informers/networking/v1beta1/destinationrule/filtered"
+)
+
+func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
+
+	logger := logging.FromContext(ctx).Desugar()
+
+	var globalResync func()
+
+	store := config.NewStore(ctx, func(name string, value interface{}) {
+		if globalResync == nil {
+			return
+		}
+
+		logger.Info("Config store changed",
+			zap.String("name", name),
+			zap.Any("value", value.(*config.Config)))
+
+		globalResync()
+	})
+	store.WatchConfigs(cmw)
+
+	// Get a filtered informer for destination rules
+	drInformer := istionetworkinginformer.Get(ctx, IstioResourceSelector)
+
+	ic := istioclientset.Get(ctx)
+	serviceInformer := serviceinformer.Get(ctx).Informer()
+
+	r := &Reconciler{
+		IstioClient:           ic,
+		DestinationRuleLister: drInformer.Lister(),
+		GetConfig: func(ctx context.Context, svc *corev1.Service) *config.Config {
+			return config.Load(store)
+		},
+	}
+
+	impl := service.NewImpl(ctx, r, func(impl *controller.Impl) controller.Options {
+		return controller.Options{
+			SkipStatusUpdates: true,
+			PromoteFilterFunc: filterServices(ctx),
+		}
+	})
+
+	globalResync = func() {
+		impl.FilteredGlobalResync(filterServices(ctx), serviceInformer)
+	}
+
+	r.Tracker = impl.Tracker
+
+	serviceInformer.AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: filterServices(ctx),
+		Handler: controller.HandleAll(controller.EnsureTypeMeta(
+			impl.Enqueue,
+			schema.GroupVersionKind{
+				Group:   corev1.SchemeGroupVersion.Group,
+				Version: corev1.SchemeGroupVersion.Version,
+				Kind:    "Service",
+			},
+		)),
+	})
+
+	// Notify the tracker that a destination rule we're tracking changed.
+	drInformer.Informer().AddEventHandler(controller.HandleAll(controller.EnsureTypeMeta(
+		impl.Tracker.OnChanged,
+		schema.GroupVersionKind{
+			Group:   istionetworking.SchemeGroupVersion.Group,
+			Version: istionetworking.SchemeGroupVersion.Version,
+			Kind:    "DestinationRule",
+		},
+	)))
+
+	return impl
+}
+
+func filterServices(ctx context.Context) func(obj interface{}) bool {
+	logger := logging.FromContext(ctx).Desugar()
+
+	return func(obj interface{}) bool {
+		imcLabels := map[string]string{
+			"messaging.knative.dev/role": "in-memory-channel",
+		}
+		kcLabels := map[string]string{
+			"messaging.knative.dev/role": "kafka-channel",
+		}
+
+		svc, ok := obj.(*corev1.Service)
+		if !ok {
+			return false
+		}
+
+		logger.Debug("Filtering Service",
+			zap.String("namespace", svc.GetNamespace()),
+			zap.String("name", svc.GetName()),
+			zap.Any("labels", svc.GetLabels()),
+		)
+
+		if svc.Spec.ExternalName == "" {
```

Comment:
when is this the case? Any

Reply:
any comment (sorry, the browser froze...)

Reply:
This is double checking that we're not queue-ing services which aren't what we expect, mostly robustness